File: frontend/de/unika/ipd/grgen/ir/expr/graph/IsInEdgesFromIndexAccessFromToExpr.java
/*
 * GrGen: graph rewrite generator tool -- release GrGen.NET 7.1
 * Copyright (C) 2003-2025 Universitaet Karlsruhe, Institut fuer Programmstrukturen und Datenorganisation, LS Goos; and free programmers
 * licensed under LGPL v3 (see LICENSE.txt included in the packaging of this file)
 * www.grgen.net
 */

package de.unika.ipd.grgen.ir.expr.graph;

import de.unika.ipd.grgen.ir.*;
import de.unika.ipd.grgen.ir.expr.Expression;
import de.unika.ipd.grgen.ir.pattern.IndexAccessOrdering;
import de.unika.ipd.grgen.ir.type.Type;

public class IsInEdgesFromIndexAccessFromToExpr extends EdgesFromIndexAccessExpr
{
	private final Expression candidateExpr;
	private final IndexAccessOrdering indexAccess;

	public IsInEdgesFromIndexAccessFromToExpr(Expression candidateExpr, IndexAccessOrdering indexAccess, Type type)
	{
		super(indexAccess.index, type);
		this.candidateExpr = candidateExpr;
		this.indexAccess = indexAccess;
	}

	public Expression getCandidateExpr()
	{
		return candidateExpr;
	}

	public IndexAccessOrdering getIndexAccessOrdering()
	{
		return indexAccess;
	}

	/** @see de.unika.ipd.grgen.ir.expr.Expression#collectNeededEntities() */
	@Override
	public void collectNeededEntities(NeededEntities needs)
	{
		needs.needsGraph();
		candidateExpr.collectNeededEntities(needs);
		indexAccess.collectNeededEntities(needs);
	}
}

File: frontend/de/unika/ipd/grgen/ir/expr/graph/IsInEdgesFromIndexAccessFromToExpr.cs
/*
 * GrGen: graph rewrite generator tool -- release GrGen.NET 8.1
 * Copyright (C) 2003-2026 Universitaet Karlsruhe, Institut fuer Programmstrukturen und Datenorganisation, LS Goos; and free programmers
 * licensed under LGPL v3, some components/parts use different licenses (see LICENSE.txt included in the packaging of this file)
 * www.grgen.de / www.grgen.net
 */

namespace de.unika.ipd.grgen.ir.expr.graph
{
	using de.unika.ipd.grgen.ir;
	using Expression = de.unika.ipd.grgen.ir.expr.Expression;
	using IndexAccessOrdering = de.unika.ipd.grgen.ir.pattern.IndexAccessOrdering;
	using Type = de.unika.ipd.grgen.ir.type.Type;

	public class IsInEdgesFromIndexAccessFromToExpr : EdgesFromIndexAccessExpr
	{
		private readonly Expression candidateExpr;
		private readonly IndexAccessOrdering indexAccess;

		public IsInEdgesFromIndexAccessFromToExpr(Expression candidateExpr, IndexAccessOrdering indexAccess, Type type)
			: base(indexAccess.index, type)
		{
			this.candidateExpr = candidateExpr;
			this.indexAccess = indexAccess;
		}

		public virtual Expression CandidateExpr
		{
			get
			{
				return candidateExpr;
			}
		}

		public virtual IndexAccessOrdering IndexAccessOrdering
		{
			get
			{
				return indexAccess;
			}
		}

		/// <seealso cref="de.unika.ipd.grgen.ir.expr.Expression.collectNeededEntities() "/>
		public override void CollectNeededEntities(NeededEntities needs)
		{
			needs.NeedsGraph();
			candidateExpr.CollectNeededEntities(needs);
			indexAccess.CollectNeededEntities(needs);
		}
	}

}
